go/tao/tpm2_tao.go

```diff
@@ -648,23 +656,92 @@ func (tt *TPM2Tao) Unseal(sealed []byte) (data []byte, policy string, err error)
 }
 
 func (s *TPM2Tao) InitCounter(label string, c int64) error {
-	fmt.Printf("tpm2.InitCounter\n") // REMOVE
-	return errors.New("InitCounter for tpm2 not implemented")
+	fmt.Printf("TPM2Tao.InitCounter\n")
+	// TODO(jlm): Change this?
+	if uint32(s.nvHandle) != 0 {
+		return nil
+	}
+	// TODO: make this more general?
+	s.nvHandle = tpm2.Handle(1000)
+	s.authString = "01020304"
+
+	return tpm2.InitCounter(s.rw, s.nvHandle, s.authString)
 }
 
 func (s *TPM2Tao) GetCounter(label string) (int64, error) {
-	fmt.Printf("tpm2.GetCounter\n") // REMOVE
-	return int64(0), errors.New("GetCounter for tpm2 not implemented")
+	fmt.Printf("TPM2Tao.GetCounter\n")
+	_ = s.InitCounter(label, int64(0))
+	return tpm2.GetCounter(s.rw, s.nvHandle, s.authString)
 }
 
+// Note:  Tpm2 counters work differently from other counters.  On startup, you
+// can't read a counter value before you initialize it which you do by incrementing
+// it.  What we do is use the (countvalue+1)/2 as the counter in RollbackSeal and Unseal.
+// When you seal, // if the current counter is odd, you bump it twice and use the 
+// value (countvalue+1)/2 in the counter slot.  If the counter is even, you bump by 1.
+// You also need to reseal the tpm keys when you startup since you may shutdown
+// before a RollbackSeal and your key will bump by two and give the wrong counter value.
+// Programmers need to know that the value returned by GetCounter is thus different from
+// the value in the sealed Rollback blob.
+
 func (s *TPM2Tao) RollbackProtectedSeal(label string, data []byte, policy string) ([]byte, error) {
-	fmt.Printf("tpm2.RollbackProtectedSeal\n") // REMOVE
-	return nil, errors.New("RollbackProtectedSeal for tpm2 not implemented")
+	fmt.Printf("tpm2tao.GetRollbackProtectedSeal\n") // REMOVE
+	_ = s.InitCounter(label, int64(0))
+	c, err := tpm2.GetCounter(s.rw, s.nvHandle, s.authString)
+	if err != nil {
+		return nil, err
+	}
+	err = tpm2.IncrementNv(s.rw, s.nvHandle, s.authString)
+	if err != nil {
+		return nil, err
+	}
+	if c % 2 == 0 {
```

Comment:
Not sure if I'm missing something but it appears to be this check should be c % 2 == 1
Say the nvram counter was say 2 before calling this function. As it is, it jumps to 5 (recall the init counter bumps it up by one too) as a result of the call which is not what we want.